core/src/main/java/org/infinispan/reactive/publisher/impl/LocalClusterPublisherManagerImpl.java

```diff
@@ -0,0 +1,112 @@
+package org.infinispan.reactive.publisher.impl;
+
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.CompletionStage;
+import java.util.function.Function;
+
+import org.infinispan.commons.util.IntSet;
+import org.infinispan.commons.util.IntSets;
+import org.infinispan.configuration.cache.Configuration;
+import org.infinispan.configuration.cache.StoreConfiguration;
+import org.infinispan.container.entries.CacheEntry;
+import org.infinispan.context.InvocationContext;
+import org.infinispan.factories.annotations.Inject;
+import org.infinispan.factories.annotations.Start;
+import org.infinispan.factories.scopes.Scope;
+import org.infinispan.factories.scopes.Scopes;
+import org.reactivestreams.Publisher;
+
+import io.reactivex.Flowable;
+import io.reactivex.processors.UnicastProcessor;
+
+@Scope(Scopes.NAMED_CACHE)
+public class LocalClusterPublisherManagerImpl<K, V> implements ClusterPublisherManager<K, V> {
+   @Inject LocalPublisherManager<K, V> localPublisherManager;
+   @Inject Configuration cacheConfiguration;
+
+   private int maxSegment;
+
+   @Start
+   public void start() {
+      // If there is a segmented store then we need to use segments
+      if (cacheConfiguration.clustering().cacheMode().needsStateTransfer() ||
+            cacheConfiguration.persistence().stores().stream().filter(StoreConfiguration::segmented).findFirst().isPresent()) {
+         maxSegment = cacheConfiguration.clustering().hash().numSegments();
+      } else {
+         maxSegment = 1;
+      }
+   }
+
+   static <K> Flowable<K> keyPublisherFromContext(InvocationContext ctx, Set<K> keysToInclude) {
+      UnicastProcessor<K> unicastProcessor = UnicastProcessor.create(ctx.lookedUpEntriesCount());
+      ctx.forEachValue((o, cacheEntry) -> {
+         K key = (K) o;
+         if (keysToInclude == null || keysToInclude.contains(key)) {
+            unicastProcessor.onNext(key);
```

Comment:
if keysToInclude is null, you cast null ? I don't get this ...

Reply:
nvm, I read this in the wrong way!!!!

Reply:
if keys to include is null in the first place or empty, we don't need to do this logic everytime. Could this be refactored or it doesn't change much